File: pkg/dart2cpp/sample/dart/test_all_functions.dart
// 测试所有函数相关表达式的转换

// 1. 普通函数定义
int add(int a, int b) {
  return a + b;
}

int subtract(int a, int b) {
  return a - b;
}

// 2. Lambda 函数表达式
var multiply = (int a, int b) => a * b;

// 3. 函数作为参数传递
int calculate(int a, int b, Function operation) {
  return operation(a, b);
}

// 4. 函数引用（Tear-off）
Function getAddFunction() {
  return add;
}

// 5. 高阶函数
Function createMultiplier(int factor) {
  return (int value) => value * factor;
}

// 6. 匿名函数
void testAnonymousFunction() {
  var result = calculate(10, 5, (int a, int b) => a + b);
  print('匿名函数结果: $result');
}

// 7. 函数变量赋值
Function? operation;

void testFunctionAssignment() {
  operation = add;
  var result = operation!(10, 5);
  print('函数变量赋值结果: $result');
}

// 8. 函数数组
List<Function> operations = [add, subtract, multiply];

void testFunctionList() {
  var result = operations[0](10, 5);
  print('函数数组结果: $result');
}

// 主函数
void main() {
  // 测试普通函数
  var result1 = calculate(10, 5, add);
  print('10 + 5 = $result1');
  
  // 测试 Lambda 函数
  var result2 = calculate(10, 5, multiply);
  print('10 * 5 = $result2');
  
  // 测试函数引用
  var addFunc = getAddFunction();
  var result3 = addFunc(10, 5);
  print('函数引用结果: $result3');
  
  // 测试高阶函数
  var multiplier = createMultiplier(3);
  var result4 = multiplier(5);
  print('高阶函数结果: $result4');
  
  // 测试匿名函数
  testAnonymousFunction();
  
  // 测试函数变量赋值
  testFunctionAssignment();
  
  // 测试函数数组
  testFunctionList();
}

File: pkg/dart2cpp/sample/cpp_generated/test_all_functions.cpp
#include "dart2cpp.h"

// 工具宏定义

Int add(Int a, Int b);
Int subtract(Int a, Int b);
Int calculate(Int a, Int b, ObjectPtr<Function> operation);
ObjectPtr<Function> getAddFunction();
ObjectPtr<Function> createMultiplier(Int factor);
Nullable testAnonymousFunction();
Nullable testFunctionAssignment();
Nullable testFunctionList();
Int add(Int a, Int b) {
  return a->operator_add(b);
}

Int subtract(Int a, Int b) {
  return a->operator_sub(b);
}

Int calculate(Int a, Int b, ObjectPtr<Function> operation) {
  return dart_cast<Int>(operation->call(a, b));
}

ObjectPtr<Function> getAddFunction() {
  return makeFunction(&add);
}

ObjectPtr<Function> createMultiplier(Int _factor) {
  ObjectPtr<_ValueBox<Int>> factor(new _ValueBox<Int>(_factor));
  return makeFunction([&, factor](Int value) mutable { return value->operator_mul((*factor)); });
}

Nullable testAnonymousFunction() {
  auto result = calculate(dart_int(10), dart_int(5), makeFunction([&](Int a, Int b) { return a->operator_add(b); }));
dart_print(dart_string("匿名函数结果: ") + (result).toString());
return Void;
}

Nullable testFunctionAssignment() {
  operation = makeFunction(&add);
Any result = operation->call(dart_int(10), dart_int(5));
dart_print(dart_string("函数变量赋值结果: ") + (result).toString());
return Void;
}

Nullable testFunctionList() {
  Any result = operations->operator_index(dart_int(0))->call(dart_int(10), dart_int(5));
dart_print(dart_string("函数数组结果: ") + (result).toString());
return Void;
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    auto result1 = calculate(dart_int(10), dart_int(5), makeFunction(&add));
dart_print(dart_string("10 + 5 = ") + (result1).toString());
auto result2 = calculate(dart_int(10), dart_int(5), multiply);
dart_print(dart_string("10 * 5 = ") + (result2).toString());
auto addFunc = getAddFunction();
Any result3 = addFunc->call(dart_int(10), dart_int(5));
dart_print(dart_string("函数引用结果: ") + (result3).toString());
auto multiplier = createMultiplier(dart_int(3));
Any result4 = multiplier->call(dart_int(5));
dart_print(dart_string("高阶函数结果: ") + (result4).toString());
testAnonymousFunction();
testFunctionAssignment();
testFunctionList();
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
